File: js/src/maplestory/api/dto/union/unionArtifactCrystalDto.ts
import { UnionArtifactCrystalBody } from '../../response/union/unionArtifactBody';

/**
 * 유니온 아티팩트 크리스탈 정보
 */
class UnionArtifactCrystalDto {
  /**
   * 아티팩트 크리스탈 명
   */
  name: string;

  /**
   * 능력치 유효 여부 (0:유효, 1:유효하지 않음)
   */
  validityFlag: string;

  /**
   * 능력치 유효 기간 (KST)
   */
  dateExpire: Date | null = null;

  /**
   * 능력치 유효 기간 만료 여부
   */
  isExpired: boolean | null = null;

  /**
   * 아티팩트 크리스탈 등급
   */
  level: number;

  /**
   * 아티팩트 크리스탈 첫 번째 옵션 명
   */
  crystalOptionName1: string;

  /**
   * 아티팩트 크리스탈 두 번째 옵션 명
   */
  crystalOptionName2: string;

  /**
   * 아티팩트 크리스탈 세 번째 옵션 명
   */
  crystalOptionName3: string;

  constructor(obj: UnionArtifactCrystalBody) {
    const {
      name,
      validity_flag,
      date_expire,
      level,
      crystal_option_name_1,
      crystal_option_name_2,
      crystal_option_name_3,
    } = obj;

    this.name = name;
    this.validityFlag = validity_flag;
    this.level = level;
    this.crystalOptionName1 = crystal_option_name_1;
    this.crystalOptionName2 = crystal_option_name_2;
    this.crystalOptionName3 = crystal_option_name_3;

    if (date_expire === 'expired') {
      this.isExpired = true;
    } else if (typeof date_expire === 'string') {
      this.dateExpire = date_expire
        ? new Date(date_expire)
        : null;
    }
  }
}

export { UnionArtifactCrystalDto };

File: java/src/main/java/dev/spiralmoon/maplestory/api/kms/dto/union/UnionArtifactCrystalDTO.java
package dev.spiralmoon.maplestory.api.kms.dto.union;

import com.google.gson.annotations.SerializedName;
import dev.spiralmoon.maplestory.api.Utils;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.ToString;

import java.time.LocalDateTime;
import java.time.ZonedDateTime;

/**
 * 유니온 아티팩트 크리스탈 정보
 */
@AllArgsConstructor
@Data
@ToString
public class UnionArtifactCrystalDTO implements dev.spiralmoon.maplestory.api.common.dto.union.UnionArtifactCrystalDTO {

    /**
     * 아티팩트 효과 명
     */
    @SerializedName("name")
    private String name;

    /**
     * 능력치 유효 여부 (0:유효, 1:유효하지 않음)
     */
    @SerializedName("validity_flag")
    private String validityFlag;

    /**
     * 능력치 유효 기간 (KST)
     */
    @SerializedName("date_expire")
    private String dateExpire;

    /**
     * 아티팩트 크리스탈 등급
     */
    @SerializedName("level")
    private int level;

    /**
     * 아티팩트 크리스탈 첫 번째 옵션 명
     */
    @SerializedName("crystal_option_name_1")
    private String crystalOptionName1;

    /**
     * 아티팩트 크리스탈 두 번째 옵션 명
     */
    @SerializedName("crystal_option_name_2")
    private String crystalOptionName2;

    /**
     * 아티팩트 크리스탈 세 번째 옵션 명
     */
    @SerializedName("crystal_option_name_3")
    private String crystalOptionName3;

    public ZonedDateTime getDateExpire() {
        if (this.dateExpire != null && !"expired".equals(this.dateExpire)) {
            return Utils.toZonedDateTime(this.dateExpire);
        } else {
            return null;
        }
    }

    /**
     * 능력치 유효 기간 만료 여부
     */
    public Boolean isExpired() {

        if (this.dateExpire == null) {
            return null;
        }

        return "expired".equals(this.dateExpire);
    }
}
